### ISSUE TITLE
Symbolic execution fails for infinite sequences in some circumstances

### ISSUE THREAD

**Issue by ddfreeman** [GaloisInc/saw-script#1703]
For reference, the below uses Cryptol 2.13.0.99 and SAW 0.9.0.99. Using the SAW Python Module, I was attempting to prove equivalence between a C function and a Cryptol one:

```c
uint32_t weird(uint32_t cv[8], uint8_t i) {
    return cv[i % 8];
}
```

```haskell
weird : [8][32] -> [8] -> [32]
weird cv i = k
    where
        k = cvs @ i
        cvs = cv # cvs
```

The idea is that I can simulate `cvs` infinitely repeating `cv` in the Cryptol implementation by taking the modulus of the index in the C implementation, since C does not support infinite sequences.

I used the following Python script to do so:

```python
class Contract_weird(Contract):
    def specification(self) -> None:
        cv = self.fresh_var(array_ty(8, i32), "cv")
        cv_p = self.alloc(array_ty(8, i32), points_to = cv, read_only=True)
        i = self.fresh_var(i8, "i")

        self.execute_func(cv_p, i)
        self.returns(cry_f("weird {cv} {i}"))
```

However, this contract failed to verify:

```
[12:09:18.830] Subgoal failed: weird internal: error: in _SAW_verify_prestate SAWServer
Literal equality postcondition

[12:09:18.830] SolverStats {solverStatsSolvers = fromList ["W4 ->z3"], solverStatsGoalSize = 990}
[12:09:18.831] ----------Counterexample----------
[12:09:18.831]   cv0: [4294967295, 0, 0, 0, 0, 0, 0, 0]
[12:09:18.831]   i0: 128
[12:09:18.831] ----------------------------------
```

So, I put that input through C and Cryptol:

```c
int main() {
    uint32_t cv[8] = {4294967295, 0, 0, 0, 0, 0, 0, 0};
    uint32_t i = 128;
    printf("0x%x\n", weird(cv, i));
}

```
```
clang weird.c -o weird.out ; ./weird.out
0xffffffff

weird> weird [4294967295, 0, 0, 0, 0, 0, 0, 0] 128
0xffffffff
```
The counterexample, erm, does not seem to be a counterexample. After some playing around, I discovered a few interesting things. First, if I hard-code a value for i, the proof succeeds:

```python
class Contract_weird(Contract):
    def specification(self) -> None:
        cv = self.fresh_var(array_ty(8, i32), "cv")
        cv_p = self.alloc(array_ty(8, i32), points_to = cv, read_only=True)
        i = self.fresh_var(i8, "i")

        self.execute_func(cv_p, cry_f("(128 : [8])"))
        self.returns(cry_f("weird {cv} 128"))

```
```
Proof succeeded! weird
✅  Verified: lemma_Contract_weird (defined at weird.py:36)
```

I'm not really an expert on symbolic execution or formal verification, but this seems to suggest that something is amiss with the symbolic execution.

Further, if `i` is a palindrome, the proof also succeeds:
```
class Contract_weird(Contract):
    def specification(self) -> None:
        cv = self.fresh_var(array_ty(8, i32), "cv")
        cv_p = self.alloc(array_ty(8, i32), points_to = cv, read_only=True)
        i = self.fresh_var(i8, "i")

        self.precondition_f("{i} == reverse {i}")

        self.execute_func(cv_p, i)
        self.returns(cry_f("weird {cv} {i}"))
```
```
Proof succeeded! weird
✅  Verified: lemma_Contract_weird (defined at weird.py:36)
```

I'm not sure what to make of this one, but it was interesting, so I felt like I should include it.

**Comment by weaversa** [comment]
If we change `weird` to:

```
weird : [8][32] -> [8] -> [32]
weird cv i = k
    where
        k = cv @ (i % 8)
```

the proof passes as well. So this looks like something going wrong with, as you say, infinite sequences. The palindrome thing is very strange.

**Comment by RyanGlScott** [comment]
It appears that using What4 is important to triggering this bug. If I change from a What4-based Z3 to an SBV-based one:

```diff
-    weird_ov = llvm_verify(mod, "weird", Contract_weird(), script=ProofScript([UseProver(Z3([]))]))
+    weird_ov = llvm_verify(mod, "weird", Contract_weird(), script=ProofScript([UseProver(Z3_SBV([]))]))
```

Then the proof succeeds.

**Comment by RyanGlScott** [comment]
Fixed in #1773.
